data/transactions/logic/eval.go

```diff
@@ -4025,48 +4019,70 @@ func (cx *EvalContext) accountReference(account stackValue) (basics.Address, uin
 	if err != nil {
 		return addr, 0, err
 	}
+
 	idx, err := cx.txn.Txn.IndexByAddress(addr, cx.txn.Txn.Sender)
+	if err == nil {
+		return addr, idx, nil
+	}
+
+	ok := cx.availableAccount(addr)
+	if !ok {
+		return addr, 0, err
+	}
+	return addr, uint64(len(cx.txn.Txn.Accounts) + 1), nil
+}
 
-	invalidIndex := uint64(len(cx.txn.Txn.Accounts) + 1)
+func (cx *EvalContext) availableAccount(addr basics.Address) bool {
 	// Allow an address for an app that was created in group
-	if err != nil && cx.version >= createdResourcesVersion {
-		for _, appID := range cx.available.apps {
+	if cx.version >= createdResourcesVersion {
+		for _, appID := range cx.available.createdApps {
 			createdAddress := cx.getApplicationAddress(appID)
 			if addr == createdAddress {
-				return addr, invalidIndex, nil
+				return true
 			}
 		}
 	}
 
+	// or some other txn mentioned it
+	if cx.version >= resourceSharingVersion {
+		if _, ok := cx.available.sharedAccounts[addr]; ok {
+			return true
+		}
+	}
+
 	// Allow an address for an app that was provided in the foreign apps array.
-	if err != nil && cx.version >= appAddressAvailableVersion {
+	if cx.version >= appAddressAvailableVersion {
 		for _, appID := range cx.txn.Txn.ForeignApps {
 			foreignAddress := cx.getApplicationAddress(appID)
 			if addr == foreignAddress {
-				return addr, invalidIndex, nil
+				return true
 			}
 		}
 	}
 
-	// this app's address is also allowed
-	if err != nil {
-		appAddr := cx.getApplicationAddress(cx.appID)
-		if appAddr == addr {
-			return addr, invalidIndex, nil
-		}
+	if cx.getApplicationAddress(cx.appID) == addr {
+		return true
 	}
 
-	return addr, idx, err
+	return false
 }
 
 func (cx *EvalContext) mutableAccountReference(account stackValue) (basics.Address, uint64, error) {
 	addr, accountIdx, err := cx.accountReference(account)
-	if err == nil && accountIdx > uint64(len(cx.txn.Txn.Accounts)) {
+	if err != nil {
+		return basics.Address{}, 0, err
+	}
+	if accountIdx > uint64(len(cx.txn.Txn.Accounts)) {
 		// There was no error, but accountReference has signaled that accountIdx
 		// is not for mutable ops (because it can't encode it in EvalDelta)
-		// This also tells us that account.address() will work.
-		addr, _ := account.address()
-		err = fmt.Errorf("invalid Account reference for mutation %s", addr)
+		if cx.version < resourceSharingVersion {
+			return basics.Address{}, 0, fmt.Errorf("invalid Account reference for mutation %s", addr)
+		}
+		// fall through, which means that starting in v9, the accountIdx
+		// returned can be > len(tx.Accounts). It will end up getting passed to
+		// GetLocal, which seems bad, since GetLocal can record that index. But
+		// that record is only done in very old consenus versions. At that point
+		// v9 did not exist.
```

Comment:
I agree, my comment alludes to danger, but doesn't explain it. It's not that bad.  How about:
```
		// fall through, which means that starting in v9, the accountIdx
		// returned can be > len(tx.Accounts). It will end up getting passed to
		// GetLocal, which can record that index in order to produce old-style
		// EDS. But those EDs are only made in old consenus versions - at that
		// point v9 did not exist, so no backward incompatible change occurs.
```

What is the thing you want tested? We do test that the "mutation" error still occurs before v9 `TestSelfMutateV8`

Reply:
Yes, a test demonstrating the issue and/or showing that local_get via `mutableAccountReference` could lead to the situation described (leaking index into apply data).

Reply:
Thank you for clarifying this - the updated comment summarizes this well. Resolved